Improve checked_* arithmetic

eschnett opened the pull request:
Add `checked_neg`, equivalent to `checked_abs`.
Implement work-arounds for currently disabled Int128 and UInt128 checked math.
Add many more tests.

eschnett committed:
Improve checked_* arithmetic

Add `checked_neg`, equivalent to `checked_abs`.
Implement work-arounds for currently disabled Int128 and UInt128 checked math.
Add many more tests.
Add documentation.

Restrict checked_abs to Int8,...,Int128, instead of handling all Signed types.
Clean up documentation of checked_abs.

nalimilan commented on base/int.jl:
```diff
@@ -552,10 +559,57 @@ end
+    x$y >= 0 && x$r < 0 && throw(OverflowError())
+    r
+end
+function checked_sub(x::Int128, y::Int128)
```
Missing line break here an below.

StefanKarpinski commented:
That's a delightfully thorough PR :-)

vtjnash commented on base/int.jl:
```diff
@@ -552,10 +559,57 @@ end
+    # x + y
+    r = x + y
+    # x and y have the same sign, and the result has a different sign
+    x$y >= 0 && x$r < 0 && throw(OverflowError())
```
any idea how that compares efficiency wise to the more literal transformation of the comment `(x > 0 == y > 0) != (x > 0 == r > 0)`?

(we also may need to be aware that this comparison on the result may cause undefined behavior in llvm -- just like C -- although i think currently we aren't setting the NSW flag)

eschnett replied:
If you replace my `&&` by `&` (why doesn't LLVM do this automatically? why!?!), then your suggestion leads to code with about three times as many machine instructions on x86-64.

eschnett replied:
Regarding undefined behaviour: We can replace the `+` with an explicit non-NSW intrinsic `wrapping_add`. Wrap-around arithmetic is convenient to have anyway.

vtjnash replied:
i'm assuming, you added parentheses when you changed `&&` to `&`? otherwise the precedence rules would rewrite that expression to `x$y >= 0 < 0` => false and should have deleted the entire thing

vtjnash commented on base/int.jl:
```diff
@@ -79,6 +79,13 @@ end
 checked_abs(x::Unsigned) = abs(x)
 
+function checked_neg{T<:Signed}(x::T)
+    x == typemin(T) && throw(OverflowError())
```
this isn't true of all Signed types, only the twos-complement representation. for example, it wouldn't be true of IEEE float (if those were a julia subtype of Signed), ones-complement, or sign-magnitude.

eschnett replied:
This mirrors existing code in `checked_abs`.

If you prefer, we can replace `Signed` by `Union{Int8,...,Int128}`, leaving the general case undefined. `BigInt` is already handled separately.

eschnett commented:
Addresses the issues, and added documentation for `checked_neg`.

eschnett commented:
Regarding efficiency:
```Julia
f1(x,y,r) = x$y>=0 && x$r<0
f2(x,y,r) = (x$y>=0) & (x$r<0)
f3(x,y,r) = ((x>=0)==(y>=0) != ((x>=0) == (r>=0)))
```
Yepp, I added parentheses around the `&` code, but I got the parentheses around your expression wrong (see above). Corrected:
```Julia
f3(x,y,r) = ((x>=0)==(y>=0)) != ((x>=0)==(r>=0))
```
your version leads to code identical to mine.

ScottPJones commented:
Might want to note that the definition of `checked_neg` depends on `Signed` only being used for 2's complement numbers (`BigInt`s are sign/magnitude, they cannot overflow with `-`, but for some reason, they are not a subtype of `Signed`, but rather directly of `Integer`.

eschnett commented:
@ScottPJones Yes; that's also the case for the existing `checked_abs`. I think my force-push deleted the respective discussion thread. As a possible solution, I offered using `Union{Int8,...,Int128}` instead of `Signed`, since `BigInt` is already handled separately.

ScottPJones commented:
Yes, I'd noticed that when `checked_abs` was added, but got too busy to try to contribute further to that discussion.  I'm adding `checked_*` functions for decimal floating point, and want to make sure that all cases are handled consistently.  Nice to see more checked arithmetic support, esp. the fixes for 128-bit arithmetic!

eschnett commented:
@ScottPJones I've tightened the signatures of `checked_abs` and `checked_neg` to only handle known integer types instead of all signed types.

tkelman commented:
It isn't immediately clear to me that all of the deleted tests are exactly subsumed by the newly added tests here, would it take more than a few seconds of runtim to leave in more of the existing tests?

eschnett commented:
All the deleted tests should be subsumed, but maybe not exactly. No harm in keeping them.

tkelman commented on base/int.jl:
```diff
@@ -43,42 +43,55 @@ copysign(x::Signed, y::Float32) = copysign(x, reinterpret(Int32,y))
+
 """
-    Base.checked_abs(x::Signed)
+    Base.checked_neg(x)
```
add signature to rst then rerun genstdlib.jl if you want the docstring to show up there

eschnett commented:
Updated signature, added documentation for the other `checked_*` functions.

tkelman commented:
lgtm
